src/signature.c

```diff
@@ -0,0 +1,179 @@
+    size_t size = 0;
+    size_t capacity = 10;
+    uint16_t* pattern = (uint16_t*)malloc(capacity * sizeof(uint16_t));
+    if (!pattern) {
+        free(signature_copy);
+        return NULL;
+    }
+
+    while (token != NULL) {
+        if (size >= capacity) {
+            capacity *= 2;
+            uint16_t* temp = (uint16_t*)realloc(pattern, capacity * sizeof(uint16_t));
+            if (!temp) {
+                free(pattern);
+                free(signature_copy);
+                return NULL;
+            }
+            pattern = temp;
+        }
+
+        if (strcmp(token, "??") == 0) {
+            pattern[size] = 0xFF00; // Set the upper byte to 1 to indicate ignoring this byte
+        } else {
+            pattern[size] = strtol(token, NULL, 16);
+        }
+        size++;
+        token = strtok(NULL, " ");
+    }
+
+    free(signature_copy);
+    *pattern_size = size;
+    return pattern;
+}
+
+bool validate_process_memory(pid_t pid, uintptr_t address, void* buffer, size_t size)
+{
+    struct iovec local_iov = { buffer, size };
+    struct iovec remote_iov = { (void*)address, size };
+    ssize_t nread = process_vm_readv(pid, &local_iov, 1, &remote_iov, 1, 0);
+
+    return nread == size;
+}
+
+int find_signature(lua_State* L)
+{
+    pid_t p_pid = process.pid;
+    const char* signature = lua_tostring(L, 1);
+    int offset = lua_tointeger(L, 2); // Get the offset as an integer directly
+
+    size_t pattern_length;
+    uint16_t* pattern = convert_signature(signature, &pattern_length);
+    if (!pattern) {
+        lua_pushinteger(L, 0);
+        return 1;
+    }
+
+    int regions_count = 0;
+    MemoryRegion* regions = get_memory_regions(p_pid, &regions_count);
+    if (!regions) {
+        free(pattern);
+        lua_pushinteger(L, 0);
+        return 1;
+    }
+
+    for (int i = 0; i < regions_count; i++) {
+        MemoryRegion region = regions[i];
+        ssize_t region_size = region.end - region.start;
+        uint8_t* buffer = (uint8_t*)malloc(region_size);
+        if (!buffer) {
+            free(pattern);
+            free(regions);
+            lua_pushinteger(L, 0);
+            return 1;
+        }
+
+        if (!validate_process_memory(p_pid, region.start, buffer, region_size)) {
+            free(buffer);
+            continue; // Continue to next region
+        }
+
+        for (size_t j = 0; j <= region_size - pattern_length; ++j) {
+            if (match_pattern(buffer + j, pattern, pattern_length)) {
+                uintptr_t result = region.start + j + offset; // Apply the offset here
+                char hex_str[20]; // Buffer to hold the hexadecimal string representation
+                sprintf(hex_str, "%lx", result); // Convert result to hexadecimal string
+
+                free(buffer);
+                free(pattern);
+                free(regions);
+
+                lua_pushstring(L, hex_str); // Push the hexadecimal string onto the Lua stack
```

Comment:
Im kinda confused, the function returns a string that contains the address in hexadecimal, but the result in the example is used as an argument for readAddress which accepts numbers, not strings.

Reply:
> Im kinda confused, the function returns a string that contains the address in hexadecimal, but the result in the example is used as an argument for readAddress which accepts numbers, not strings. Is it casting it to a number by some miracle?

Yeah, Lua seems to automatically convert the hex string. So I didn't deem it necessary to convert it myself.

In the Jet Set Radio autosplitter, I replaced one of the hex numbers with a string.
```lua
    current.isLoading = readAddress('bool', "0x58FAAC")
    current.newGame = readAddress('bool', 0x75A278)
    current.rankingScreen = readAddress('bool', 0x58FB1C)
    current.bossGraffiti = readAddress('int', 0x55D2B8)
```

and it still worked fine.